### ISSUE TITLE
vAmigaWeb

### ISSUE THREAD

**Issue by mithrendal** [nicodex/amiga-ocs-cpubltro#1]
cool stuff 😀

when putting it into https://vAmigaWeb.github.io 

I see during animation every second frame or so that it changes from this 

<img width="796" alt="image" src="https://github.com/user-attachments/assets/5a77b39b-3c6f-499e-9b2b-c6f6ee84c0aa">

to this 

<img width="796" alt="image" src="https://github.com/user-attachments/assets/50dd6d64-d2a9-420b-b6e3-40aa83cfc110">

is this effect

<img width="60" alt="image" src="https://github.com/user-attachments/assets/399d0562-1f05-4ac9-9347-e357b6f63faa">

normal and intended ?

BTW: I activated the activity monitor with this symbol 

<img width="87" alt="image" src="https://github.com/user-attachments/assets/f636439b-c99e-4acc-a9ab-8be15615cb71">

and indeed no write access. I see only bitplane DMA

<img width="795" alt="image" src="https://github.com/user-attachments/assets/a6f15fd6-c562-4ee9-b4c2-12ce24ae948c">

**Comment by nicodex** [comment]
Before blaming vAmiga I'll try to grab the screen with a RGBtoHDMI on real hardware, as soon as my spare time allows.
My interlaced handling just might be wrong...

At least I noticed no problems on real hardware (A500 without DRAM, CIAs, and Paula). I added a photo from a live hardware demo on a retro meeting last weekend, see https://github.com/nicodex/amiga-ocs-cpubltro/commit/f2a00badf2b02e38b870ec8d37bc436e0aeb7e6c

**Comment by mithrendal** [comment]
yeah, maybe it looks so because vAmigaWeb uses no shader to emulate CRT blurring

**Comment by mithrendal** [comment]
hmm its not the shader CRT thing, I think. 

In @dirkwhoffmann 's  original `vAmiga for Mac` which supports various kinds of fancy settings to emulate a CRT monitor 

<img width="592" alt="image" src="https://github.com/user-attachments/assets/c70d98b9-7d7d-47c4-8112-825087d96d15">


I played with the settings. 

Above settings makes it look like this below. 



![image](https://github.com/user-attachments/assets/c177ae69-10f4-44b6-a99d-5f9b2ca0b6c7)

Not always but every n_th frame... it looks like that


I suppose the shadow effect for the red squares on the ball should be always like this maybe ?

![image](https://github.com/user-attachments/assets/92f2109c-9e58-4a15-ae85-7f1e5208827d)

aboves picture which looks correct to me, shows up every m_th frame.


>The remaining ROM space allows us to include 11/256K animation frames.

where 
   0< m ... n < 11 

or in words some of the frames render like a beautiful amiga ball 😀 others show some artefacts...

**Comment by nicodex** [comment]
The 256K ROM has 11 images, the 512K ROM has 22 images. No RAM, no fancy stuff, just fullscreen image data. For the live demo I hacked the animation logic together with the assumtion that a PAL frame starts with a long field, followed by a short field. The negated LOF flag is or-ed into the field counter, which is shifted right as image index. This should result in the same image for the first/long and the second/short field. However, this is interlaced - after the long field you see half of two images. So this 'problem' is expected.
As soon as I have the time, I will have a look at it. Maybe the effect can be avoided... but this is a hardware related demo in the first place :-)

**Comment by mithrendal** [comment]
>However, this is interlaced - after the long field you see half of two images. So this 'problem' is expected.

Ah I see it now too 🤓. The "artefacts" are because of interlace mode... which mixes half pictures together. During animation we see then on every second frame one half picture which is still old ... 

Impressive. The Amiga needs less ram for this demo (0KB) then the VCS2600 with its small 128 Bytes of RAM. Really cool 😎.

**Comment by nicodex** [comment]
A PAL Amiga has a 28.37516 MHz crystal oscillator. Divide this by 8 to get the Amiga CCK of 3,546895 MHz (Amiga CCK * 5 / 4 = 4.43361875 MHz PAL color subcarrier). A PAL frame (two fields) has 625 lines. Every line has 227 Amiga CCKs. Depending on the LOF bit the Amiga counts/generates 313 or 312 lines. Enabling the LACE bit in the chipset just toggles the LOF bit. Therefore we have three different field frequencies on PAL Amigas:
- 2 / (227 * (313 + 312) * 3.546895 MHz) = 50.000282 Hz (interlaced)
- 2 / (227 * (312 + 312) * 3.546895 MHz) = 50.080411 Hz (non-laced short)
- 2 / (227 * (313 + 313) * 3.546895 MHz) = 49.920409 Hz (non-laced long)
(nice to know: the AmigaOS explicitly sets the LOF bit for non-laced modes)

So the interlaced mode is closer to the PAL spec (that's the main reason why I'm using it).

However, there is no such thing as a non-laced/progressive mode for the 15 kHz Amiga display modes. We always have to generate/fill the color data for every odd and even field. In theory it does not matter when the animation is advanced. Switching the image every field would be fine - but the animation would be really fast. So I decided to display a single image for two fields. It even does not matter if the next image starts in the even or the odd field, only every other display line will be updated. It makes a difference for display devices that convert from 50i into 25p - only for them it is important that the first and the second field contain the same image.

If all my assumptions are correct (and LOF bit is set for the first PAL field), this issue will be a `wontfix` :-)

**Comment by mithrendal** [comment]
Yes its not an issue 😅. Lets close. Brilliant. Thanks for in depth explanation.

**Comment by dirkwhoffmann** [comment]
I've added the fingerprints of the two ROMs (the 512 KB and the 256 KB versions) to vAmiga's ROM database so they can be installed simply via drag and drop in the next release:

<img width="1019" alt="Bildschirmfoto 2024-11-21 um 16 24 50" src="https://github.com/user-attachments/assets/cad85948-c1b7-41f2-9957-11bcd8e41274">

Please let me know if the Roms should be displayed with a different text in the ROM config panel. 

Very cool demo, btw. Based on what the activity monitors tell us, the machine looks dead but apparently, it is still alive:

<img width="1019" alt="Bildschirmfoto 2024-11-21 um 16 25 11" src="https://github.com/user-attachments/assets/fb9ecdd4-61e7-4d2e-bbaf-d5e3166523c9">

**Comment by nicodex** [comment]
> I've added the fingerprints of the two ROMs

Many thanks for making it more accessible :)

> Please let me know if the Roms should be displayed with a different text in the ROM config panel.

I have no preferences. The name 'cpubltro' is somewhat unwieldy, but that's how the project started :)

**Comment by nicodex** [comment]
PAL field/line numbering is odd (pun intended).
Reopened until I have a better understanding of the Amiga interlaced mode versus PAL signals. Recent discussion on Discord indicates that I'm wrong about the PAL field order.

**Comment by dirkwhoffmann** [comment]
> Recent discussion on Discord indicates that I'm wrong about the PAL field order.

Is the PAL field order really relevant? At some point, you need to switch to a texture showing the ball at a different position, and as your demo runs in interlace mode, the other half of the image will still show the ball where it was before. Thus, if I am not missing something fundamental, the stripes should be unavoidable, shouldn't they?

**Comment by nicodex** [comment]
> Is the PAL field order really relevant?

Not really, except for my innermost need to understand how it works. I'm still quite new to the Amiga hardware. Back in the days I did not care how a CRT works, but now I'm interested in breaking down how the Agnus and the Denise work together to generate the signals. So I'll hack some code together and see how it looks on my ancient Amiga 1081 monitor, an old TV, and a modern TFT - just to actually see the differences, if there are any :)

**Comment by nicodex** [comment]
Most/all projects agree that a PAL interlaced frame starts with a _short_ field.
e.g. https://github.com/nonarkitten/amiga_replacement_project/blob/ad6b61d7763eb3edadf4277d13b12bdeebf6f583/denise/ice/denise.v#L109

Since I want to know the initial and the reset state of a real Agnus for another project anyway, I'll keep this open for reference/documentation purposes. But for this project I will not create a new release (because it does not change anything for my hardware setup with a CRT).
